0000-overhaul-qiskit-experiments.md

```diff
@@ -0,0 +1,638 @@
+
+I also scrutinized the performance of the current framework with a statistical profiler.
+Since the percentage of function detection varied significantly from run to run, I only show qualitative results here.
+I found following subroutines tend to frequently show up in the frame graph.
+
+#### Initialization of `ExperimentData` sub-container
+
+Because composite analysis needs to instantiate new data container to run component analyses,
+calling the constructor of this expensive class consumes significant time in heavily batched experiment.
+Especially, acquisition of a service object and qiskit version information is the expensive part.
+Since such information is not necessary for inner (child) data container, we can replace `ExperimentData`
+with much lightweight data container to alleviate this overhead. See the following discussion for details.
+
+#### Marginalization of the result count
+
+When parallelize the experimental circuits, the measured count dictionary is keyed on the
+merged bit string of all measurements in sub experiments.
+The max length of the dictionary scales with O(n2) with qubit size, and actual size 
+may increase with shot number. Since the marginalization function is already implemented in Rust,
+we need to reduce number of call to this function.
+
+#### Creation of curve-fit figures
+
+When we run a curve analysis, the analysis will generate a matplotlib figure to interact with human experimentalists.
+This subroutine is also expensive. Fortunately we can set `plot=False` in the analysis options
+if figure is not necessary. 
+Experimentalist will be able to load raw data points from the experiment service at later time (not supported as of QE 0.5). 
+
+#### Curve fitting
+
+This is not significantly slow but the overhead is visible. It also may depend on the analysis.
+Some analysis provide multiple initial guesses for the fit model parameters,
+and the fitter is called many times for each guess values.
+
+
+## User Benefit
+
+A user or system agent who heavily rely on the parallel or batch execution feature will dramatically benefit from this overhaul. 
+These users typically be a responsible for the system deployment or maybe a researcher who benchmarks a novel gate with larger application circuits. 
+This kind of work requires users to run a chain of calibration and monitoring experiments at a scale of the quantum processor.
+If experiments are conducted in the dedicated queue mode or with a [runtime in a session](https://qiskit.org/ecosystem/ibm-runtime/how_to/run_session.html), 
+the significant overhead in the local analysis task may also deteriorate the machine efficiency of the remote quantum computing system.
+
+With the proposed overhaul plan, we aim to achieve the T1 analysis wall time of < 5 sec at nq=100 either with a high-end laptop or workstation (sorry we don't have proper benchmark suite for QE!).
```

Comment:
I've started something here: https://github.com/eliarbel/qiskit-experiments-benchmarks. We should discuss if and how we want to leverage this or something like this for formalizing a benchmark framework for QE

Reply:
I don't have access to the repo, but I remember you showed a handmade benchmark suite. If we want to continuously monitor the performance, I guess we need to setup some bare metal server. @mtreinish knows how to setup the benchmak.

Reply:
Ideally we should have a bare metal server and something like ASV. Let's have a separate discussion about this aspect.